src/pm.c

```diff
@@ -0,0 +1,472 @@
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+#define PM_COUNTER	"counter"
+#define PM_GAUGE	"gauge"
+#define PM_HISTOGRAM	"histogram"
+
+#define PM_LEN			128
+#define PM_LABEL_MAX		5
+#define PM_LABEL_VAL_MAX	5
+#define PM_METRIC_MAX		30
+#define PM_HIST_BIN_MAX		10
+
+typedef struct {
+	char		name[PM_LEN];
+	char		help[PM_LEN];
+	char		type[PM_LEN];
+	char		label[PM_LABEL_MAX][PM_LEN];
+	uint32_t	n_label;
+	double		bin[PM_HIST_BIN_MAX];
+	uint32_t	n_bin;
+	struct {
+		char	lblval[PM_LABEL_MAX][PM_LEN];
+		union {
+			double	 	counter;
+			double		gauge;
+			struct {
+				double		sum;
+				uint64_t	cnt[PM_HIST_BIN_MAX];
+				uint64_t	inf;
+			} histogram;
+		};
+	} s[PM_LABEL_VAL_MAX];
+	uint32_t	n_store;
+} pm_metric;
+
+static int		 pm_metric_get_store(pm_metric *m, va_list *);
+static pm_metric	 metrics[PM_METRIC_MAX] = {{{0}}};
+
+int
+pm_metric_get_store(pm_metric *m, va_list *ap1)
+{
+	va_list		ap;
+	uint32_t	i, j;
+	int		idx = -1;
+	uint32_t	ret;
+	uint8_t		match;
+	const char	*label;
+	char		lblval[PM_LABEL_MAX][PM_LEN] = {{0}};
+
+	va_copy(ap, *ap1);
+	for (i = 0; i < m->n_label; i++) {
+		label = va_arg(ap, const char *);
+
+		for (j = 0; j < m->n_label; j++) {
+			if (strcmp(label, m->label[j]) == 0)
+				break;
+		}
+		ret = snprintf(lblval[j], sizeof(lblval[j]), "\"%s\"",
+			va_arg(ap, const char *));
+		if (ret >= sizeof(lblval[j])) goto out;
+	}
+	va_end(ap);
+
+	match = 0;
+	for (i = 0; i < PM_LABEL_VAL_MAX; i++) {
+		if (m->s[i].lblval[0][0] == '\0') {
+			if (idx == -1)
+				idx = i;
+			continue;
+		}
+		for (j = 0; j < m->n_label; j++) {
+			if (strcmp(m->s[i].lblval[j], lblval[j]) != 0)
+				break;
+		}
+		if (j == m->n_label) {
+			match = 1;
+			idx = i;
+			break;
+		}
+	}
+
+	if (match == 0 && idx != -1) {
+		for (i = 0; i < m->n_label; i++) {
+			snprintf(m->s[idx].lblval[i],
+				sizeof(m->s[idx].lblval[i]), "%s", lblval[i]);
+			if (ret >= sizeof(lblval[j])) goto out;
+		}
+		m->n_store++;
+	}
+out:
+	return (idx);
+}
+
+
+pm_metric *
+pm_metric_add(const char *name, const char *help, const char *type, const uint32_t n_label, ...)
+{
+	pm_metric	*m = NULL;
+	va_list		 args;
+	uint32_t	 i, j;
+	uint32_t	 ret = 0;
+
+	for (i = 0; i < PM_METRIC_MAX; i++)
+		if (metrics[i].name[0] == '\0')
+			break;
+
+	if (i >= PM_METRIC_MAX)
+		goto out;
+
+	va_start(args, n_label);
+	for (j = 0; j < n_label && j < PM_LABEL_MAX; j++) {
+		ret = snprintf(metrics[i].label[j], sizeof(metrics[i].label[j]),
+			"%s", va_arg(args, const char *));
+		if (ret >= sizeof(metrics[i].label[j])) goto out;
+	}
+	va_end(args);
+
+	ret = snprintf(metrics[i].help, sizeof(metrics[i].help), "%s", help);
+	if (ret >= sizeof(metrics[i].help)) goto out;
+
+	ret = snprintf(metrics[i].type, sizeof(metrics[i].type), "%s", type);
+	if (ret > sizeof(metrics[i].type)) goto out;
+
+	ret = snprintf(metrics[i].name, sizeof(metrics[i].name), "%s", name);
+	if (ret > sizeof(metrics[i].name)) {
+		metrics[i].name[0] = '\0';
```

Comment:
Could this default zero value string be defined above as a variable and reference instead of hardcoded like this? I see that it's also checked on line 69. Might just help a bit towards readability.

Reply:
hmmm I think I will use strlen() on the string strlen(metric[i].name]) == 0, I guess that will be clearer, what do you think ?